cantools/database/dataelement.py

```diff
@@ -0,0 +1,192 @@
+            if not isinstance(scale, list) or not isinstance(offset, list):
+                raise ValueError(
+                    "Params scale and offset need to be of type list "
+                    "if segment boundaries are defined."
+                )
+
+            self._initialize_segment_limits(segment_boundaries)
+
+    def _initialize_segment_limits(self, segment_boundaries: PiecewiseSegment) -> None:
+        def convert(v, o, f):
+            return v * f + o
+
+        self.segment_boundaries: PiecewiseSegment = []
+        last_phys_max = None
+        for i, segment in enumerate(segment_boundaries):
+            self.segment_boundaries_raw.append(segment)
+            start, end = segment
+            params = [self._offset[i], self._scale[i]]  # type: ignore
+            scaled_segment = (convert(start, *params), convert(end, *params))
+            self.segment_boundaries_scaled.append(scaled_segment)
+            if last_phys_max is None:
+                last_phys_max = scaled_segment[1]
+                continue
+
+            if last_phys_max >= scaled_segment[0]:
+                logger.warning(
+                    f"Piecewise linear type: {self.name} has overlapping segments! "
+                    f"Segment {i} starts at phys val {scaled_segment[0]} "
+                    f"but one of the prev segments ended at {last_phys_max}. "
+                    "Encoding might be ambiguous."
+                )
+
+            last_phys_max = max(scaled_segment[1], last_phys_max)
+
+    def choice_string_to_number(self, string: str) -> int:
+        if self.choices is None:
+            raise ValueError(f"Signal {self.name} has no choices.")
+
+        for choice_number, choice_value in self.choices.items():
+            if str(choice_value) == str(string):
+                return choice_number
+
+        raise KeyError(f"Choice {string} not found in data element {self.name}.")
+
+    def _get_offset_scaling_from_list(self, val: float, segments: PiecewiseSegment) -> Tuple[float, float]:
+        for i, segment in enumerate(segments):
+            start, end = segment
+            if start <= val <= end:
+                return self._offset[i], self._scale[i]  # type: ignore
+        else:
+            err_text = [f"{start} <= x <= {end}" for start, end in segments]
+            raise ValueError(
+                f"Value {val} is not in ranges: \n {' OR '.join(err_text)}"
+            )
+
+    def get_offset_scaling_from_raw(
+        self, raw_val: Optional[float] = None
+    ) -> Tuple[float, float]:
+        """Get the applicable offset and scaling for the given raw value.
+
+        If data type only defines one set of offset/scaling then
+        the `raw_val` param can be omitted
+        """
+        if raw_val is None or len(self.segment_boundaries_raw) == 0:
+            try:
+                return self._offset[0], self._scale[0]  # type: ignore
+            except TypeError:
+                return self.offset, self.scale
```

Comment:
```suggestion
        if raw_val is None:
                return (0.0, 1.0)
        elif not self.segment_boundaries_raw:
                return (self.offset, self.scale)
```

Reply:
hmm this code does not exactly achieve the same results. If we have a (not piecewise) linear type the `offset` and `scale` may be different from the defaults but will also be independent from the input value so `raw_val` might be `None` and we still want the actual stored factors.

Since we dont know whether `offset` and `scale` are iterable or not here, I figured the most "pythonic" way to handle the situation would be to just try and catch the exception if it fails but it's certainly not my favourite solution, so if you have any other idea I would still be open :)